### PR TITLE
Complete implementation of the protocol

### PR DESCRIPTION
**Pull request by felipediel**
## Notice
This PR got too big. I am dividing it into smaller parts and merging separately to avoid regressions.. Then I will continue here. Don't waste time looking at the code, I will change it later.

## Breaking changes
- exceptions.check_error() now takes an integer as argument
- device.send_packet() now returns the decrypted message
- device.hello() now returns a dictionary and doesn't update device information automatically. Use device.update() instead.

They are unlikely to impact older applications. Most of this is recent / for internal use.

## Prevent data corruption
- Check the length of the data packets
- Check the sum of the encrypted data packets
- Check the information type of the data packets
- Check the connection ID before decrypting a message
- Do all protocol-level checks inside device.send_packet() to retry with the same packet number (prevents duplicate commands when the data is corrupted)

## Multi-threading without locks
- Track responses by packet number and remove the lock of device.send_packet()
- Add a lock to device.auth() in order to make it thread-safe

## More features
- Add device.update() method to fetch data in a safe way (it uses encryption, so it's better than hello(), which should be used to start the communication only)
- Add support for the LB27 R1 bulb

## For developers
- Create a DeviceCore abstraction to segregate interfaces and make it easier to add support for new devices
- Create BlockSizeHandler classes to help adding support for devices that require special headers for communication
- Create an EncryptionHandler class to abstract encryption/decryption and make it easier to cope to breaking changes in the protocol
- Create a device.send_cmd() abstraction to improve code readability
- Allow sending public requests and public responses with device.send_packet()
- Allow any Sequence type in device.send_packet() and device.send_cmd()
- Create a State class to help packing and unpacking state for transport

### CODE REVIEW

**Comment by kylemd** [comment]
Seems to work fine here:

Screenshot_20210226-073446_Termux

**Comment by felipediel** [comment]
This last error concerns me. Did it work before?

**Comment by kylemd** [comment]
Nope, these are the lightbulbs referenced in #522.

File: broadlink/device.py [hunk 1]
```diff
@@ -231,56 +278,198 @@ def set_lock(self, state: bool) -> None:
...
+                        -4012,
+                        "Device control ID error",
+                        f"Expected {self.id} and received {dev_ctrl_id}",
+                    )
+
+                payload = self.decrypt(payload)
+                nom_p_checksum = int.from_bytes(resp[0x34:0x36], "little")
+                real_p_checksum = (
+                    p_checksum if err_code else sum(payload, 0xBEAF) & 0xFFFF
+                )
```
Comment (felipediel) [inline broadlink/device.py:463]: There was a breaking change in the protocol at some point. The new firmware works like this:
```python3
real_p_checksum = sum(payload, 0xBEAF) & 0xFFFF
```
I need more time to investigate. I think it is reasonable to log and skip this error in v0.17.0.

**Comment by felipediel** [comment]
Hi @kylemd. Does it work with the LB27 now?

**Comment by kylemd** [comment]
Screenshot_20210301-071222_Termux

Yup, works just fine!

**Comment by felipediel** [comment]
Great, thank you! I need to clean this up now, we'll need a final test later.

**Comment by kylemd** [comment]
Just trying to test this out in HA, still getting the device not supported error. 

Gave the instructions from [your post here](https://github.com/mjg59/python-broadlink/issues/480#issuecomment-752183830)

```
docker exec -it homeassistant /bin/bash
pip3 install git+https://github.com/felipediel/python-broadlink.git@the-protocol --force-reinstall --no-deps
```

Am I missing something here? Thought I might mention it in case something has broken.

**Comment by felipediel** [comment]
We still need to add support for bulbs there. It will take some time.

**Comment by kylemd** [comment]
Copy that, thanks!

**Comment by felipediel** [comment]
abstractions

**Comment by felipediel** [comment]
Ok, now we have a decent implementation of the protocol. It is not yet complete, but it is better than before.

Hey @kylemd, mind testing the LB27 again? I hope the code survived all these changes 😆

**Comment by kylemd** [comment]
Yep, working fine here!

**Comment by felipediel** [comment]
Thank you!

**Comment by MichielioZ** [comment]
@felipediel Man, seems like you've been busy ! Don't want to rush anything, just trying to get an estimate on when the code for the LB27 R1 bulbs  might be available in HA ? (I know it's hard to predict such things)

**Comment by felipediel** [comment]
It will take a long time. We have more than 3000 lines of code in the queue.

**Comment by felipediel** [comment]
It is being exhausting to implement the protocol without any documentation and with so many people asking for the release. I will make a cut so I can continue calmly. Some features will not be available in the next version.

**Comment by kylemd** [comment]
Don't let them get you down - the amount of work you've put into this project is beyond impressive! Looking forward to what you put out next 😃👌

**Comment by felipediel** [comment]
Thank you for the encouraging words. But I will have to cut. People don't understand that we skipped a version. They want the new version within the deadline of the previous. That's impossible.

The truth is that v0.16.1 should have happened, it is not healthy for the project to make so many changes at once. I will get back here to bring your bulbs, don't worry.

**Comment by kylemd** [comment]
> I will get back here to bring your bulbs, don't worry.

Haha, no rush! They all work with Google Home already, just not HA, so they're sloooooowwww. 
Following the project has been fascinating and I'm enjoying helping where I can, so notifications are on 😄

**Comment by felipediel** [comment]
I'm afraid I will never find the time to finish this. Closing, I will reopen if I do.

**Pull request closed without merging**